File: app/src/main/java/com/coinninja/coinkeeper/model/helpers/DaoSessionManager.java
package com.coinninja.coinkeeper.model.helpers;

import androidx.annotation.NonNull;

import com.coinninja.coinkeeper.model.db.AccountDao;
import com.coinninja.coinkeeper.model.db.Address;
import com.coinninja.coinkeeper.model.db.AddressDao;
import com.coinninja.coinkeeper.model.db.BroadcastBtcInviteDao;
import com.coinninja.coinkeeper.model.db.DaoMaster;
import com.coinninja.coinkeeper.model.db.DaoSession;
import com.coinninja.coinkeeper.model.db.DropbitMeIdentity;
import com.coinninja.coinkeeper.model.db.DropbitMeIdentityDao;
import com.coinninja.coinkeeper.model.db.ExternalNotificationDao;
import com.coinninja.coinkeeper.model.db.FundingStat;
import com.coinninja.coinkeeper.model.db.FundingStatDao;
import com.coinninja.coinkeeper.model.db.InternalNotificationDao;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummaryDao;
import com.coinninja.coinkeeper.model.db.TargetStat;
import com.coinninja.coinkeeper.model.db.TargetStatDao;
import com.coinninja.coinkeeper.model.db.TransactionNotification;
import com.coinninja.coinkeeper.model.db.TransactionNotificationDao;
import com.coinninja.coinkeeper.model.db.TransactionSummary;
import com.coinninja.coinkeeper.model.db.TransactionSummaryDao;
import com.coinninja.coinkeeper.model.db.TransactionsInvitesSummary;
import com.coinninja.coinkeeper.model.db.TransactionsInvitesSummaryDao;
import com.coinninja.coinkeeper.model.db.User;
import com.coinninja.coinkeeper.model.db.UserDao;
import com.coinninja.coinkeeper.model.db.UserIdentity;
import com.coinninja.coinkeeper.model.db.UserIdentityDao;
import com.coinninja.coinkeeper.model.db.Wallet;
import com.coinninja.coinkeeper.model.db.WalletDao;
import com.coinninja.coinkeeper.model.db.WordDao;
import com.coinninja.coinkeeper.service.client.model.GsonAddress;

import org.greenrobot.greendao.database.Database;
import org.greenrobot.greendao.query.QueryBuilder;

import javax.inject.Singleton;

@Singleton
public class DaoSessionManager {
    private DaoSession session;
    private DaoMaster daoMaster;

    public DaoSessionManager(DaoMaster daoMaster) {
        this.daoMaster = daoMaster;
    }

    public void clear() {
        session.clear();
    }

    public DaoSessionManager connect() {
        session = daoMaster.newSession();
        return this;
    }

    @Deprecated
    public DaoSession getDaoSession() {
        return session;
    }

    public void resetAll() {
        clear();
        dropAllTables();
        createAllTables();
        connect();
    }

    public void runRaw(String query) {
        session.getDatabase().execSQL(query);
    }

    // USER
    public UserDao getUserDao() {
        return session.getUserDao();
    }

    // WALLET
    public WalletDao getWalletDao() {
        return session.getWalletDao();
    }

    public QueryBuilder<Wallet> qeuryForWallet() {
        return getWalletDao().queryBuilder();
    }

    public Wallet createWallet() {
        resetAll();
        User user = new User();
        long id = getUserDao().insert(user);
        Wallet wallet = new Wallet();
        wallet.setUserId(id);
        getWalletDao().insert(wallet);
        return wallet;
    }

    // WORD
    public WordDao getWordDao() {
        return session.getWordDao();
    }

    // ADDRESS
    public AddressDao getAddressDao() {
        return session.getAddressDao();
    }

    // TARGET STAT
    public TargetStatDao getTargetStatDao() {
        return session.getTargetStatDao();
    }

    public long insert(TargetStat targetStat) {
        long id = getTargetStatDao().insert(targetStat);
        targetStat.setId(id);
        targetStat.__setDaoSession(session);
        return id;
    }

    public TargetStat newTargetStat() {
        return new TargetStat();
    }

    // FUNDING STAT
    public FundingStatDao getFundingStatDao() {
        return session.getFundingStatDao();
    }

    public FundingStat newFundingStat() {
        return new FundingStat();
    }

    public long insert(FundingStat fundingStat) {
        long id = session.insert(fundingStat);
        fundingStat.setId(id);
        fundingStat.__setDaoSession(session);
        return id;
    }

    // TRANSACTION INVITE SUMMARY
    public TransactionsInvitesSummaryDao getTransactionsInvitesSummaryDao() {
        return session.getTransactionsInvitesSummaryDao();
    }

    public TransactionsInvitesSummary newTransactionInviteSummary() {
        return new TransactionsInvitesSummary();
    }

    public long insert(TransactionsInvitesSummary summary) {
        long id = getTransactionsInvitesSummaryDao().insert(summary);
        summary.setId(id);
        summary.__setDaoSession(session);
        return id;
    }

    // TRANSACTION
    public TransactionSummaryDao getTransactionSummaryDao() {
        return session.getTransactionSummaryDao();
    }

    public long insert(TransactionSummary transaction) {
        long id = getTransactionSummaryDao().insert(transaction);
        transaction.setId(id);
        transaction.__setDaoSession(session);
        return id;
    }

    public void attach(TransactionSummary transactionSummary) {
        transactionSummary.__setDaoSession(session);
    }

    public TransactionSummary newTransactionSummary() {
        return new TransactionSummary();
    }

    // TRANSACTION NOTIFICATION
    public long insert(TransactionNotification notification) {
        return getTransactionNotificationDao().insert(notification);
    }

    public TransactionNotification newTransactionNotification() {
        return new TransactionNotification();
    }


    public TransactionNotificationDao getTransactionNotificationDao() {
        return session.getTransactionNotificationDao();
    }

    // DROPBIT
    public InviteTransactionSummaryDao getInviteTransactionSummaryDao() {
        return session.getInviteTransactionSummaryDao();
    }

    public long insert(InviteTransactionSummary inviteTransactionSummary) {
        long id = session.getInviteTransactionSummaryDao().insert(inviteTransactionSummary);
        inviteTransactionSummary.__setDaoSession(session);
        inviteTransactionSummary.setId(id);
        return id;
    }

    public InviteTransactionSummary newInviteTransactionSummary() {
        return new InviteTransactionSummary();
    }

    // ACCOUNT
    public AccountDao getAccountDao() {
        return session.getAccountDao();
    }

    // INTERNAL NOTIFICATION
    public InternalNotificationDao getInternalNotificationDao() {
        return session.getInternalNotificationDao();
    }

    public ExternalNotificationDao getExternalNotificationDao() {
        return session.getExternalNotificationDao();
    }

    public BroadcastBtcInviteDao getBroadcastBtcInviteDao() {
        return session.getBroadcastBtcInviteDao();
    }

    // Dropbit Me Identity
    public DropbitMeIdentityDao getDropbitMeIdentityDao() {
        return session.getDropbitMeIdentityDao();
    }

    public DropbitMeIdentity newDropbitMeIdentity() {
        return new DropbitMeIdentity();
    }

    public long insert(DropbitMeIdentity dropbitMeIdentity) {
        long id = getDropbitMeIdentityDao().insert(dropbitMeIdentity);
        dropbitMeIdentity.__setDaoSession(session);
        dropbitMeIdentity.setId(id);
        return id;
    }

    // UserIdentity
    public UserIdentity newUserIdentity() {
        return new UserIdentity();
    }

    public UserIdentityDao getUserIdentityDao() {
        return session.getUserIdentityDao();
    }

    public long insert(UserIdentity userIdentity) {
        long id = session.insert(userIdentity);
        userIdentity.setId(id);
        userIdentity.__setDaoSession(session);
        return id;
    }

    public void clearCacheFor(TransactionSummary transaction) {
        //Clear cache of to-many relations
        transaction.resetFunder();
        transaction.resetReceiver();
    }

    @NonNull
    public Address newAddressFrom(GsonAddress gsonAddress, Wallet wallet, int changeIndex) {
        Address address = new Address();
        address.setWallet(wallet);
        address.setAddress(gsonAddress.getAddress());
        address.setChangeIndex(gsonAddress.getDerivationIndex());
        address.setChangeIndex(changeIndex);
        address.setIndex(gsonAddress.getDerivationIndex());
        session.getAddressDao().insert(address);
        return address;
    }

    private void dropAllTables() {
        Database db = session.getDatabase();
        DaoMaster.dropAllTables(db, true);
    }

    private void createAllTables() {
        Database db = session.getDatabase();
        DaoMaster.createAllTables(db, true);
    }

}

File: app/src/main/java/com/coinninja/coinkeeper/model/helpers/DaoSessionManager.kt
package com.coinninja.coinkeeper.model.helpers

import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.model.db.*
import com.coinninja.coinkeeper.service.client.model.GsonAddress
import org.greenrobot.greendao.query.QueryBuilder
import javax.inject.Singleton

@Singleton
@Mockable
class DaoSessionManager(
        internal val daoMaster: DaoMaster,
        internal val purpose: Int = 49,
        internal val coinType: Int = 0,
        internal val accountIndex: Int = 0
) {
    internal lateinit var daoSession: DaoSession

    // USER
    val userDao: UserDao get() = daoSession.userDao

    // WALLET
    val walletDao: WalletDao get() = daoSession.walletDao

    fun createWallet(): Wallet {
        resetAll()
        val user = User()
        val id = userDao.insert(user)
        val wallet = Wallet().also {
            it.userId = id
            it.purpose = purpose
            it.coinType = coinType
            it.accountIndex = accountIndex
        }
        val walletId = walletDao.insert(wallet)
        wallet.id = walletId
        wallet.__setDaoSession(daoSession)
        return wallet
    }

    // WORD
    val wordDao: WordDao get() = daoSession.wordDao

    // ADDRESS
    val addressDao: AddressDao get() = daoSession.addressDao

    fun newAddressFrom(gsonAddress: GsonAddress, wallet: Wallet, changeIndex: Int): Address {
        val address = Address()
        address.wallet = wallet
        address.address = gsonAddress.address
        address.changeIndex = gsonAddress.derivationIndex
        address.changeIndex = changeIndex
        address.index = gsonAddress.derivationIndex
        daoSession.addressDao.insert(address)
        return address
    }

    // TARGET STAT
    val targetStatDao: TargetStatDao get() = daoSession.targetStatDao

    // FUNDING STAT
    val fundingStatDao: FundingStatDao get() = daoSession.fundingStatDao

    // TRANSACTION INVITE SUMMARY
    val transactionsInvitesSummaryDao: TransactionsInvitesSummaryDao get() = daoSession.transactionsInvitesSummaryDao

    // TRANSACTION
    val transactionSummaryDao: TransactionSummaryDao get() = daoSession.transactionSummaryDao


    val transactionNotificationDao: TransactionNotificationDao get() = daoSession.transactionNotificationDao

    // DROPBIT
    val inviteTransactionSummaryDao: InviteTransactionSummaryDao get() = daoSession.inviteTransactionSummaryDao

    // ACCOUNT
    val accountDao: AccountDao get() = daoSession.accountDao

    // INTERNAL NOTIFICATION
    val internalNotificationDao: InternalNotificationDao
        get() = daoSession.internalNotificationDao

    val externalNotificationDao: ExternalNotificationDao
        get() = daoSession.externalNotificationDao

    val broadcastBtcInviteDao: BroadcastBtcInviteDao
        get() = daoSession.broadcastBtcInviteDao

    // Dropbit Me Identity
    val dropbitMeIdentityDao: DropbitMeIdentityDao get() = daoSession.dropbitMeIdentityDao

    val userIdentityDao: UserIdentityDao get() = daoSession.userIdentityDao

    fun clear() {
        daoSession.clear()
    }

    fun connect(): DaoSessionManager {
        daoSession = daoMaster.newSession()
        return this
    }

    fun resetAll() {
        clear()
        dropAllTables()
        createAllTables()
        connect()
    }

    fun runRaw(query: String) {
        daoSession.database.execSQL(query)
    }

    fun qeuryForWallet(): QueryBuilder<Wallet> {
        return walletDao.queryBuilder()
    }

    // Account
    fun insert(account: Account): Long {
        val id = accountDao.insert(account)
        account.__setDaoSession(daoSession)
        account.id = id
        return id
    }

    fun insert(targetStat: TargetStat): Long {
        val id = targetStatDao.insert(targetStat)
        targetStat.id = id
        targetStat.__setDaoSession(daoSession)
        return id
    }

    fun newTargetStat(): TargetStat {
        return TargetStat()
    }

    fun newFundingStat(): FundingStat {
        return FundingStat()
    }

    fun insert(fundingStat: FundingStat): Long {
        val id = daoSession.insert(fundingStat)
        fundingStat.id = id
        fundingStat.__setDaoSession(daoSession)
        return id
    }

    fun newTransactionInviteSummary(): TransactionsInvitesSummary {
        return TransactionsInvitesSummary()
    }

    fun insert(summary: TransactionsInvitesSummary): Long {
        val id = transactionsInvitesSummaryDao.insert(summary)
        summary.id = id
        summary.__setDaoSession(daoSession)
        return id
    }

    fun insert(transaction: TransactionSummary): Long {
        val id = transactionSummaryDao.insert(transaction)
        transaction.id = id
        transaction.__setDaoSession(daoSession)
        return id
    }

    fun attach(transactionSummary: TransactionSummary) {
        transactionSummary.__setDaoSession(daoSession)
    }

    fun newTransactionSummary(): TransactionSummary {
        return TransactionSummary()
    }

    // TRANSACTION NOTIFICATION
    fun insert(notification: TransactionNotification): Long {
        return transactionNotificationDao.insert(notification)
    }

    fun newTransactionNotification(): TransactionNotification {
        return TransactionNotification()
    }

    fun insert(inviteTransactionSummary: InviteTransactionSummary): Long {
        val id = daoSession.inviteTransactionSummaryDao.insert(inviteTransactionSummary)
        inviteTransactionSummary.__setDaoSession(daoSession)
        inviteTransactionSummary.id = id
        return id
    }

    fun newInviteTransactionSummary(): InviteTransactionSummary {
        return InviteTransactionSummary()
    }

    fun newDropbitMeIdentity(): DropbitMeIdentity {
        return DropbitMeIdentity()
    }

    fun insert(dropbitMeIdentity: DropbitMeIdentity): Long {
        val id = dropbitMeIdentityDao.insert(dropbitMeIdentity)
        dropbitMeIdentity.__setDaoSession(daoSession)
        dropbitMeIdentity.id = id
        return id
    }

    // UserIdentity
    fun newUserIdentity(): UserIdentity {
        return UserIdentity()
    }

    fun insert(userIdentity: UserIdentity): Long {
        val id = daoSession.insert(userIdentity)
        userIdentity.id = id
        userIdentity.__setDaoSession(daoSession)
        return id
    }

    fun clearCacheFor(transaction: TransactionSummary) {
        //Clear cache of to-many relations
        transaction.resetFunder()
        transaction.resetReceiver()
    }

    private fun dropAllTables() {
        val db = daoSession.database
        DaoMaster.dropAllTables(db, true)
    }

    private fun createAllTables() {
        val db = daoSession.database
        DaoMaster.createAllTables(db, true)
    }

}
